src/main/java/org/junit/runners/parameterized/BlockJUnit4ClassRunnerWithParameters.java

```diff
@@ -135,7 +138,73 @@ protected void validateFields(List<Throwable> errors) {
 
     @Override
     protected Statement classBlock(RunNotifier notifier) {
-        return childrenInvoker(notifier);
+        Statement statement = childrenInvoker(notifier);
+        statement = withBeforeParams(statement);
+        statement = withAfterParams(statement);
+        return statement;
+    }
+
+    private Statement withBeforeParams(Statement statement) {
+        List<FrameworkMethod> befores = getTestClass()
+                .getAnnotatedMethods(Parameterized.BeforeParam.class);
+        return befores.isEmpty() ? statement : new RunBeforeParams(statement, befores);
+    }
+
+    private class RunBeforeParams extends Statement {
+        private final Statement next;
+        private final List<FrameworkMethod> befores;
+
+        RunBeforeParams(Statement next, List<FrameworkMethod> befores) {
+            this.next = next;
+            this.befores = befores;
+        }
+
+        @Override
+        public void evaluate() throws Throwable {
+            for (FrameworkMethod before : befores) {
+                int paramCount = before.getMethod().getParameterTypes().length;
+                before.invokeExplosively(
+                        null, paramCount == 0 ? (Object[]) null : parameters);
+            }
+            next.evaluate();
+        }
+    }
+
+    private Statement withAfterParams(Statement statement) {
+        List<FrameworkMethod> afters = getTestClass()
+                .getAnnotatedMethods(Parameterized.AfterParam.class);
+        return afters.isEmpty() ? statement : new RunAfterParams(statement, afters);
+    }
+
+    private class RunAfterParams extends Statement {
+        private final Statement next;
+        private final List<FrameworkMethod> afters;
+
+        RunAfterParams(Statement next, List<FrameworkMethod> afters) {
+            this.next = next;
+            this.afters = afters;
+        }
+
+        @Override
+        public void evaluate() throws Throwable {
+            List<Throwable> errors = new ArrayList<Throwable>();
+            try {
+                next.evaluate();
+            } catch (Throwable e) {
+                errors.add(e);
+            } finally {
+                for (FrameworkMethod each : afters) {
+                    try {
+                        int paramCount = each.getMethod().getParameterTypes().length;
+                        each.invokeExplosively(
+                                null, paramCount == 0 ? (Object[]) null : parameters);
+                    } catch (Throwable e) {
+                        errors.add(e);
+                    }
+                }
+            }
+            MultipleFailureException.assertEmpty(errors);
+        }
```

Comment:
Can we reuse `org.junit.internal.runners.statements.RunAfters` and pass `parameters` to it?